csharp/src/Ice/SlicBinaryConnection.cs

```diff
@@ -107,164 +161,233 @@ public async ValueTask CloseAsync(Exception exception, CancellationToken cancel)
-                    return (StreamId: requestId, Frame: frame, Fin: requestId == 0 || frame is IncomingResponseFrame);
+                    case FrameType.Ping:
+                    {
+                        ValueTask task = PrepareAndSendFrameAsync(FrameType.Pong, null, CancellationToken.None);
+                        _heartbeatCallback();
+                        break;
+                    }
+                    case FrameType.Pong:
+                    {
+                        // TODO: setup a timer to expect pong frame response?
+                        break;
+                    }
+                    case FrameType.Stream:
+                    case FrameType.StreamLast:
+                    {
+                        (long streamId, int streamIdSize) = data.AsReadOnlySpan().ReadVarLong();
+                        data = data.Slice(streamIdSize);
+                        IncomingFrame frame = ParseIce2Frame(data);
+                        ProtocolTrace.TraceFrame(Endpoint.Communicator, streamId, data.Count, frame);
+                        return (StreamId: streamId, Frame: frame, Fin: type == FrameType.StreamLast);
+                    }
+                    case FrameType.ResetStream:
+                    {
+                        var istr = new InputStream(data, Encoding);
+                        long streamId = istr.ReadVarLong();
+                        long reason = istr.ReadVarLong();
+                        // TODO: release resources allocated for the stream.
+                        break;
+                    }
+                    case FrameType.Close:
+                    {
+                        var istr = new InputStream(data, Encoding);
+                        long code = istr.ReadVarLong();
+                        string reason = istr.ReadString();
+                        throw new ConnectionClosedByPeerException();
+                    }
+                    default:
+                    {
+                        throw new InvalidDataException($"unexpected Slic frame with frame type `{type}'");
+                    }
                 }
             }
         }
 
         public long NewStream(bool bidirectional) => bidirectional ? ++_nextStreamId : 0;
 
-        public ValueTask ResetAsync(long streamId) =>
-            throw new NotSupportedException("ice2 transports don't support stream reset");
+        public ValueTask ResetAsync(long streamId) => PrepareAndSendFrameAsync(FrameType.ResetStream, ostr =>
+            {
+                ostr.WriteVarLong(streamId);
+                ostr.WriteVarLong(0);
+            }, CancellationToken.None);
 
-        public async ValueTask SendAsync(long streamId, OutgoingFrame frame, bool fin, CancellationToken cancel) =>
-            await SendFrameAsync(streamId, frame, cancel);
+        public async ValueTask SendAsync(long streamId, OutgoingFrame frame, bool fin, CancellationToken cancel)
+        {
+            var data = new List<ArraySegment<byte>>();
```

Comment:
Would it make sense the reserve some space in frame.Data and use this space for the Slic header (for small ice2 frames)?

Reply:
Surely if it can reduce the number of segment allocations... we could do the same for Ice1 as well. I guess we could just reserve 26 bytes. It should be good enough for:
- the Ice1 header (14 bytes) + requestID (4 bytes) or
- the Slic framing (frame type, frame size and stream ID, max 17 bytes) + Ice2 frame type and frame size (max 9 bytes).
For another PR?

Reply:
Yes, definitely for another PR.